### ISSUE TITLE
vw yarn, stuck at map 100% when training set is large

### ISSUE THREAD

**Issue by robbine** [JohnLangford/vowpal_wabbit#1039]
I tried mapscript-yarn.sh and fed it with tens of millions of training samples.
The hadoop streaming job stuck at 100% and never end.
All mappers have 100% progress bar while state remains "RUNNING"
Does anyone know how to solve this?

I runs well with small training set

**Comment by JohnLangford** [comment]
A common issue is that the number of mappers is not controlled.  Are you
controlling it to force a number of mappers which can all run
simultaneously?

You need to trick Hadoop into having a fixed-size set of mappers because
synchronous communication is required.   To do this partition the dataset
with .gz files (which Hadoop cannot break up) or run VW in a (controlled)
number of reducers.

-John

**Comment by robbine** [comment]
Some of the mappers failed with exit code 141
Error: java.lang.RuntimeException: PipeMapRed.waitOutputThreads(): subprocess failed with code 141 at org.apache.hadoop.streaming.PipeMapRed.waitOutputThreads(PipeMapRed.java:320) at

**Comment by robbine** [comment]
Some of the earlier failed mapper print out logs like this "Can't create directory /srv/data/0/yarn/local/usercache/recsys/"

**Comment by robbine** [comment]
This is a stderr log of a failed mapper
000068
100
hdfs://hz-cluster1/tmp/model3
newsrec3.photo.163.org
Starting training
final_regressor = sgd.vwmodel
Num weight bits = 20
learning rate = 0.5
initial_t = 0
power_t = 0.5
decay_learning_rate = 1
creating cache_file = temp.cache
Reading datafile = /dev/stdin
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
0.693147 0.693147            1            1.0   1.0000   0.0000      736
0.754372 0.815597            2            2.0  -1.0000   0.2315      493
0.729902 0.705432            4            4.0  -1.0000   0.0490      804
0.752557 0.775212            8            8.0   1.0000  -0.3097     1336
0.709228 0.665899           16           16.0  -1.0000  -0.2002      928
0.713395 0.717562           32           32.0  -1.0000  -0.0867      388
0.691629 0.669863           64           64.0  -1.0000  -1.0638      628
0.733293 0.774958          128          128.0   1.0000  -1.2202      850
0.711297 0.689300          256          256.0  -1.0000  -0.8662      685
0.729741 0.748184          512          512.0  -1.0000  -0.3634      697
0.722237 0.714733         1024         1024.0  -1.0000  -1.3837      820
0.733208 0.744179         2048         2048.0  -1.0000   0.2466      145
0.720030 0.706852         4096         4096.0  -1.0000  -0.4074      166
0.708118 0.696207         8192         8192.0  -1.0000   0.0658     1213
0.701159 0.694199        16384        16384.0  -1.0000  -0.9834      631
0.690789 0.680420        32768        32768.0   1.0000  -1.0445      988
connecting to 10.160.128.51 = newsrec3.photo.163.org:26543
wrote unique_id=124187
wrote total=100
wrote node=68
read ok=1
read kid_count=0
read parent_ip=172.17.2.218
read parent_port=45159
connecting to 172.17.2.218 = hadoop134.photo.163.org:26544
0.649341 0.649341        65536        65536.0  -1.0000  -2.2413      145 h
Net time taken by process = 45.193000 seconds

finished run
number of examples per pass = 4593530
passes used = 2
weighted example sum = 9187060.000000
weighted label sum = -1767496.000000
average loss = 62.040253 h
best constant = -0.389635
best constant's loss = 0.674524
total feature number = 6344128512
final_regressor = bfgs.vwmodel
enabling BFGS based optimization **without** curvature calculation
Num weight bits = 20
learning rate = 0.5
initial_t = 0
power_t = 0.5
decay_learning_rate = 1
m = 5
Allocated 56M for weights and mem
## avg. loss 	der. mag. 	d. m. cond.	 wolfe1    	wolfe2    	mix fraction	curvature 	dir. magnitude	step size 
using cache_file = temp.cache
ignoring text input in favor of cache input
num sources = 1
connecting to 10.160.128.51 = newsrec3.photo.163.org:26543
wrote unique_id=248374
wrote total=100
wrote node=68
read ok=1
read kid_count=0
read parent_ip=172.17.2.218
read parent_port=45159
connecting to 172.17.2.218 = hadoop134.photo.163.org:26544
 1 0.58032   	0.00145   	4364.62468	          	          	          	8168261037.90242	687001501696.00000	0.00000   
Maximum number of passes reached. If you want to optimize further, increase the number of passes
Net time taken by process = 32.405 seconds
 recv from parent: Connection reset by peer
log4j:WARN No appenders could be found for logger (org.apache.hadoop.metrics2.impl.MetricsSystemImpl).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.

**Comment by arielf** [comment]
Why do you think this is a `vw` issue?

**Comment by robbine** [comment]
maybe it is NOT. hadoop fs command is not allowed in the cluster I use. 
The only way to solve this is by using multiple output format.

**Comment by JohnLangford** [comment]
Sounds like it's not a vw issue.  Closing for now.
